<!DOCTYPE html>
<html lang="en" dir=" auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">
<title>An R function for MAF calculation | Mohammad Ali Nilforooshan</title>
<meta name="keywords" content="R, Quantitative Genetics">
<meta name="description" content="Here, I introduce an R function for calculating minor allele frequencies (MAF).
calcmaf &lt;- function(M, col1ID = TRUE) {
    if(col1ID) {
        maf = colMeans(M[,-1])/2
    } else {
        maf = colMeans(M)/2
    }
    maf[maf &gt; 0.5] &lt;- 1 - maf[maf &gt; 0.5]
    return(unname(maf))
}
The calcMAF function takes arguments M and col1ID.
M is the genotype data frame with genotypes coded as 0:2.
col1ID takes TRUE or FALSE.
If TRUE (default) the 1st column of M is animal ID.
Let&rsquo;s create an example genotype data frame for 10 genotypes and 20 SNPs, where the first column is animal ID.">
<meta name="author" content="Mohammad Ali Nilforooshan">
<link rel="canonical" href="https://nilforooshan.github.io/blog/posts/calcmaf/">
<link crossorigin="anonymous" href="/blog/assets/css/stylesheet.45e028aa8ce0961349adf411b013ee39406be2c0bc80d4ea3fc04555f7f4611a.css" integrity="sha256-ReAoqozglhNJrfQRsBPuOUBr4sC8gNTqP8BFVff0YRo=" rel="preload stylesheet" as="style">
<link rel="icon" href="https://nilforooshan.github.io/blog/favicon.ico">
<link rel="icon" type="image/png" sizes="16x16" href="https://nilforooshan.github.io/blog/favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="https://nilforooshan.github.io/blog/favicon-32x32.png">
<link rel="apple-touch-icon" href="https://nilforooshan.github.io/blog/apple-touch-icon.png">
<link rel="mask-icon" href="https://nilforooshan.github.io/blog/safari-pinned-tab.svg">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://nilforooshan.github.io/blog/posts/calcmaf/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --code-block-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript>
      <script async src="https://www.googletagmanager.com/gtag/js?id=G-JH0F7W96MC"></script>
      <script>
        var doNotTrack = false;
        if ( false ) {
          var dnt = (navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack);
          var doNotTrack = (dnt == "1" || dnt == "yes");
        }
        if (!doNotTrack) {
          window.dataLayer = window.dataLayer || [];
          function gtag(){dataLayer.push(arguments);}
          gtag('js', new Date());
          gtag('config', 'G-JH0F7W96MC');
        }
      </script><meta property="og:url" content="https://nilforooshan.github.io/blog/posts/calcmaf/">
  <meta property="og:site_name" content="Mohammad Ali Nilforooshan">
  <meta property="og:title" content="An R function for MAF calculation">
  <meta property="og:description" content="Here, I introduce an R function for calculating minor allele frequencies (MAF).
calcmaf &lt;- function(M, col1ID = TRUE) { if(col1ID) { maf = colMeans(M[,-1])/2 } else { maf = colMeans(M)/2 } maf[maf &gt; 0.5] &lt;- 1 - maf[maf &gt; 0.5] return(unname(maf)) } The calcMAF function takes arguments M and col1ID. M is the genotype data frame with genotypes coded as 0:2. col1ID takes TRUE or FALSE. If TRUE (default) the 1st column of M is animal ID.
Let’s create an example genotype data frame for 10 genotypes and 20 SNPs, where the first column is animal ID.">
  <meta property="og:locale" content="en">
  <meta property="og:type" content="article">
    <meta property="article:section" content="posts">
    <meta property="article:published_time" content="2025-02-28T00:00:00+00:00">
    <meta property="article:modified_time" content="2025-02-28T00:00:00+00:00">
    <meta property="article:tag" content="R">
    <meta property="article:tag" content="Quantitative Genetics">
      <meta property="og:image" content="https://nilforooshan.github.io/blog/%3Clink%20or%20path%20of%20image%20for%20opengraph,%20twitter-cards%3E">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="https://nilforooshan.github.io/blog/%3Clink%20or%20path%20of%20image%20for%20opengraph,%20twitter-cards%3E">
<meta name="twitter:title" content="An R function for MAF calculation">
<meta name="twitter:description" content="Here, I introduce an R function for calculating minor allele frequencies (MAF).
calcmaf &lt;- function(M, col1ID = TRUE) {
    if(col1ID) {
        maf = colMeans(M[,-1])/2
    } else {
        maf = colMeans(M)/2
    }
    maf[maf &gt; 0.5] &lt;- 1 - maf[maf &gt; 0.5]
    return(unname(maf))
}
The calcMAF function takes arguments M and col1ID.
M is the genotype data frame with genotypes coded as 0:2.
col1ID takes TRUE or FALSE.
If TRUE (default) the 1st column of M is animal ID.
Let&rsquo;s create an example genotype data frame for 10 genotypes and 20 SNPs, where the first column is animal ID.">


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "Posts",
      "item": "https://nilforooshan.github.io/blog/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "An R function for MAF calculation",
      "item": "https://nilforooshan.github.io/blog/posts/calcmaf/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "An R function for MAF calculation",
  "name": "An R function for MAF calculation",
  "description": "Here, I introduce an R function for calculating minor allele frequencies (MAF).\ncalcmaf \u0026lt;- function(M, col1ID = TRUE) { if(col1ID) { maf = colMeans(M[,-1])/2 } else { maf = colMeans(M)/2 } maf[maf \u0026gt; 0.5] \u0026lt;- 1 - maf[maf \u0026gt; 0.5] return(unname(maf)) } The calcMAF function takes arguments M and col1ID. M is the genotype data frame with genotypes coded as 0:2. col1ID takes TRUE or FALSE. If TRUE (default) the 1st column of M is animal ID.\nLet\u0026rsquo;s create an example genotype data frame for 10 genotypes and 20 SNPs, where the first column is animal ID.\n",
  "keywords": [
    "R", "Quantitative Genetics"
  ],
  "articleBody": "Here, I introduce an R function for calculating minor allele frequencies (MAF).\ncalcmaf \u003c- function(M, col1ID = TRUE) { if(col1ID) { maf = colMeans(M[,-1])/2 } else { maf = colMeans(M)/2 } maf[maf \u003e 0.5] \u003c- 1 - maf[maf \u003e 0.5] return(unname(maf)) } The calcMAF function takes arguments M and col1ID. M is the genotype data frame with genotypes coded as 0:2. col1ID takes TRUE or FALSE. If TRUE (default) the 1st column of M is animal ID.\nLet’s create an example genotype data frame for 10 genotypes and 20 SNPs, where the first column is animal ID.\nset.seed(995) M \u003c- matrix(c(sample(0:2, 200, replace = TRUE)), nrow = 10) (M \u003c- as.data.frame(cbind(101:110, M))) V1 V2 V3 V4 V5 V6 V7 V8 V9 V10 V11 V12 V13 V14 V15 V16 V17 V18 V19 V20 V21 1 101 2 0 2 1 0 2 1 1 1 0 1 1 2 1 0 1 2 0 1 1 2 102 1 2 0 0 1 1 1 2 2 0 0 2 2 0 1 2 2 2 1 0 3 103 0 0 2 1 0 1 2 1 0 2 2 1 1 0 0 0 0 2 0 1 4 104 1 1 2 1 0 2 2 1 1 1 0 1 2 1 0 2 2 0 2 0 5 105 0 2 1 1 2 2 2 1 1 2 1 0 2 2 2 1 2 0 1 1 6 106 0 1 2 0 0 0 2 0 0 0 0 1 0 1 0 1 1 1 1 2 7 107 2 2 2 2 0 0 1 0 0 2 1 0 2 2 1 0 2 0 0 0 8 108 2 1 2 1 1 1 0 1 0 0 2 1 0 1 1 2 2 2 0 1 9 109 2 1 0 2 0 0 0 0 0 0 0 0 0 2 0 1 0 0 2 2 10 110 2 2 0 2 0 0 1 0 0 1 0 1 1 2 0 1 1 2 0 2 Now, calculate MAF:\ncalcmaf(M) # equivalent to calcmaf(M, col1ID = TRUE) [1] 0.40 0.40 0.35 0.45 0.20 0.45 0.40 0.35 0.25 0.40 0.35 0.40 0.40 0.40 0.25 [16] 0.45 0.30 0.45 0.40 0.50 Suppose, we had a data frame, purely genotypes, where the first column was not animal ID (M[,-1]). Then:\ncalcmaf(M = M[,-1], col1ID = FALSE) [1] 0.40 0.40 0.35 0.45 0.20 0.45 0.40 0.35 0.25 0.40 0.35 0.40 0.40 0.40 0.25 [16] 0.45 0.30 0.45 0.40 0.50 ",
  "wordCount" : "432",
  "inLanguage": "en",
  "image": "https://nilforooshan.github.io/blog/%3Clink%20or%20path%20of%20image%20for%20opengraph,%20twitter-cards%3E","datePublished": "2025-02-28T00:00:00Z",
  "dateModified": "2025-02-28T00:00:00Z",
  "author":{
    "@type": "Person",
    "name": "Mohammad Ali Nilforooshan"
  },
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://nilforooshan.github.io/blog/posts/calcmaf/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Mohammad Ali Nilforooshan",
    "logo": {
      "@type": "ImageObject",
      "url": "https://nilforooshan.github.io/blog/favicon.ico"
    }
  }
}
</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.10/dist/katex.min.css" integrity="sha384-wcIxkf4k558AjM3Yz3BBFQUbk/zgIYC2R0QpeeYb+TwlBVMrlgLqwRjRtGZiK7ww" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.10/dist/katex.min.js" integrity="sha384-hIoBPJpTUs74ddyc4bFZSM1TVlQDA60VBbJS0oA934VSz82sBx1X7kSx2ATBDIyd" crossorigin="anonymous"></script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.10/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous"
  onload="renderMathInElement(document.body);"></script>

<script>
    document.addEventListener("DOMContentLoaded", function() {
        renderMathInElement(document.body, {
          delimiters: [
            {left: '$$', right: '$$', display: true},
            {left: '$', right: '$', display: false},
            {left: "\\begin{equation}", right: "\\end{equation}", display: true},
            {left: "\\begin{equation*}", right: "\\end{equation*}", display: true},
            {left: "\\begin{align}", right: "\\end{align}", display: true},
            {left: "\\begin{align*}", right: "\\end{align*}", display: true},
            {left: "\\begin{alignat}", right: "\\end{alignat}", display: true},
            {left: "\\begin{gather}", right: "\\end{gather}", display: true},
            {left: "\\begin{CD}", right: "\\end{CD}", display: true},
          ],
          throwOnError : false
        });
    });
</script>


<script async src="https://www.googletagmanager.com/gtag/js?id=G-JH0F7W96MC"></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() { dataLayer.push(arguments); }
    gtag('js', new Date());

    gtag('config', 'G-JH0F7W96MC');
</script>
</head>

<body class="" id="
    top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://nilforooshan.github.io/blog/" accesskey="h" title="Mohammad Ali Nilforooshan (Alt + H)">Mohammad Ali Nilforooshan</a>
            <div class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="18" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="18" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
            </div>
        </div>
        <ul id="menu">
            <li>
                <a href="https://nilforooshan.github.io/blog/" title="Blog">
                    <span>Blog</span>
                </a>
            </li>
            <li>
                <a href="https://nilforooshan.github.io/blog/tags/" title="Tags">
                    <span>Tags</span>
                </a>
            </li>
            <li>
                <a href="https://nilforooshan.github.io/blog/search/" title="Search (Alt &#43; /)" accesskey=/>
                    <span>Search</span>
                </a>
            </li>
            <li>
                <a href="https://nilforooshan.github.io/" title="My website">
                    <span>My website</span>&nbsp;
                    <svg fill="none" shape-rendering="geometricPrecision" stroke="currentColor" stroke-linecap="round"
                        stroke-linejoin="round" stroke-width="2.5" viewBox="0 0 24 24" height="12" width="12">
                        <path d="M18 13v6a2 2 0 01-2 2H5a2 2 0 01-2-2V8a2 2 0 012-2h6"></path>
                        <path d="M15 3h6v6"></path>
                        <path d="M10 14L21 3"></path>
                    </svg>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    
    <h1 class="post-title entry-hint-parent">
      An R function for MAF calculation
    </h1>
    <div class="post-meta"><span title='2025-02-28 00:00:00 +0000 UTC'>February 28, 2025</span>&nbsp;·&nbsp;Mohammad Ali Nilforooshan

</div>
  </header> 
  <div class="post-content"><p>Here, I introduce an R function for calculating minor allele frequencies (MAF).</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>calcmaf <span style="color:#f92672">&lt;-</span> <span style="color:#66d9ef">function</span>(M, col1ID <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>) {
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">if</span>(col1ID) {
</span></span><span style="display:flex;"><span>        maf <span style="color:#f92672">=</span> <span style="color:#a6e22e">colMeans</span>(M[,<span style="color:#ae81ff">-1</span>])<span style="color:#f92672">/</span><span style="color:#ae81ff">2</span>
</span></span><span style="display:flex;"><span>    } <span style="color:#66d9ef">else</span> {
</span></span><span style="display:flex;"><span>        maf <span style="color:#f92672">=</span> <span style="color:#a6e22e">colMeans</span>(M)<span style="color:#f92672">/</span><span style="color:#ae81ff">2</span>
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    maf[maf <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">0.5</span>] <span style="color:#f92672">&lt;-</span> <span style="color:#ae81ff">1</span> <span style="color:#f92672">-</span> maf[maf <span style="color:#f92672">&gt;</span> <span style="color:#ae81ff">0.5</span>]
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">return</span>(<span style="color:#a6e22e">unname</span>(maf))
</span></span><span style="display:flex;"><span>}
</span></span></code></pre></div><p>The <code>calcMAF</code> function takes arguments <code>M</code> and <code>col1ID</code>.
<code>M</code> is the genotype data frame with genotypes coded as <code>0:2</code>.
<code>col1ID</code> takes <code>TRUE</code> or <code>FALSE</code>.
If <code>TRUE</code> (default) the 1st column of <code>M</code> is animal ID.</p>
<p>Let&rsquo;s create an example genotype data frame for 10 genotypes and 20 SNPs, where the first column is animal ID.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#a6e22e">set.seed</span>(<span style="color:#ae81ff">995</span>)
</span></span><span style="display:flex;"><span>M <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">matrix</span>(<span style="color:#a6e22e">c</span>(<span style="color:#a6e22e">sample</span>(<span style="color:#ae81ff">0</span><span style="color:#f92672">:</span><span style="color:#ae81ff">2</span>, <span style="color:#ae81ff">200</span>, replace <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>)), nrow <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span>)
</span></span><span style="display:flex;"><span>(M <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">as.data.frame</span>(<span style="color:#a6e22e">cbind</span>(<span style="color:#ae81ff">101</span><span style="color:#f92672">:</span><span style="color:#ae81ff">110</span>, M)))
</span></span></code></pre></div><pre tabindex="0"><code>    V1 V2 V3 V4 V5 V6 V7 V8 V9 V10 V11 V12 V13 V14 V15 V16 V17 V18 V19 V20 V21
1  101  2  0  2  1  0  2  1  1   1   0   1   1   2   1   0   1   2   0   1   1
2  102  1  2  0  0  1  1  1  2   2   0   0   2   2   0   1   2   2   2   1   0
3  103  0  0  2  1  0  1  2  1   0   2   2   1   1   0   0   0   0   2   0   1
4  104  1  1  2  1  0  2  2  1   1   1   0   1   2   1   0   2   2   0   2   0
5  105  0  2  1  1  2  2  2  1   1   2   1   0   2   2   2   1   2   0   1   1
6  106  0  1  2  0  0  0  2  0   0   0   0   1   0   1   0   1   1   1   1   2
7  107  2  2  2  2  0  0  1  0   0   2   1   0   2   2   1   0   2   0   0   0
8  108  2  1  2  1  1  1  0  1   0   0   2   1   0   1   1   2   2   2   0   1
9  109  2  1  0  2  0  0  0  0   0   0   0   0   0   2   0   1   0   0   2   2
10 110  2  2  0  2  0  0  1  0   0   1   0   1   1   2   0   1   1   2   0   2
</code></pre><p>Now, calculate MAF:</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#a6e22e">calcmaf</span>(M) <span style="color:#75715e"># equivalent to calcmaf(M, col1ID = TRUE)</span>
</span></span></code></pre></div><pre tabindex="0"><code> [1] 0.40 0.40 0.35 0.45 0.20 0.45 0.40 0.35 0.25 0.40 0.35 0.40 0.40 0.40 0.25
[16] 0.45 0.30 0.45 0.40 0.50
</code></pre><p>Suppose, we had a data frame, purely genotypes, where the first column was not animal ID (<code>M[,-1]</code>). Then:</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#a6e22e">calcmaf</span>(M <span style="color:#f92672">=</span> M[,<span style="color:#ae81ff">-1</span>], col1ID <span style="color:#f92672">=</span> <span style="color:#66d9ef">FALSE</span>)
</span></span></code></pre></div><pre tabindex="0"><code> [1] 0.40 0.40 0.35 0.45 0.20 0.45 0.40 0.35 0.25 0.40 0.35 0.40 0.40 0.40 0.25
[16] 0.45 0.30 0.45 0.40 0.50
</code></pre>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://nilforooshan.github.io/blog/tags/r/">R</a></li>
      <li><a href="https://nilforooshan.github.io/blog/tags/quantitative-genetics/">Quantitative Genetics</a></li>
    </ul>
  </footer>
</article>
    </main>
    
<footer class="footer">
        <span>&copy; 2025 <a href="https://nilforooshan.github.io/blog/">Mohammad Ali Nilforooshan</a></span> · 

    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://github.com/adityatelange/hugo-PaperMod/" rel="noopener" target="_blank">PaperMod</a>
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
</body>

</html>